src/main/java/org/apache/onami/persist/EntityManagerProviderImpl.java

```diff
@@ -105,6 +112,36 @@ private EntityManager createEntityManager() {
   /**
    * {@inheritDoc}
    */
+  @Override
+  public TransactionHolder withTransaction(final TransactionFacade transactionFacade) {
+    final TransactionFacade original = transactionFacades.get();
+    transactionFacades.set(transactionFacade);
+    return new TransactionHolder() {
+      @Override
+      public TransactionFacade getTransaction() {
+        return transactionFacade;
+      }
+
+      @Override
+      public void close() throws Exception {
+        transactionFacades.set(original);
+      }
+    };
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void addPostCommitCallback(Runnable callback) {
+    TransactionFacade transactionFacade = transactionFacades.get();
+    Preconditions.checkNotNull(transactionFacade, "Cannot add a callback outside the context of a transaction");
+    transactionFacade.addPostCommitCallback(callback);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
   // @Override
```

Comment:
We do cleanup at the end of each transaction boundary instead, since that's more correct in the case where a session encompasses multiple transactions.